test/rbc/rbc_test.go

```diff
@@ -0,0 +1,240 @@
+	var ip string = "127.0.0.1"
+	var port uint16 = 8000
+
+	nodeList := make([]*mockNode, 0)
+
+	for i := 0; i < n; i++ {
+		availablePort := GetAvailablePort(port)
+		address := cleisthenes.Address{
+			Ip:   ip,
+			Port: availablePort,
+		}
+		node := newMockNode(n, f, address)
+		node.setUpServer()
+		go node.listen()
+		nodeList = append(nodeList, node)
+
+		port = availablePort + 1
+	}
+
+	for m, mNode := range nodeList {
+		// memberMap include proposer node
+		memberMap := cleisthenes.NewMemberMap()
+
+		// proposer (owner)
+		member := cleisthenes.NewMember(mNode.address.Ip, mNode.address.Port)
+		memberMap.Add(member)
+
+		// connPool has only other nodes (except me)
+		connPool := cleisthenes.NewConnectionPool()
+
+		//
+		// set client and memberMap
+		//
+		for y, yNode := range nodeList {
+			if m != y {
+				cli := cleisthenes.NewClient()
+				conn, err := cli.Dial(cleisthenes.DialOpts{
+					Addr: cleisthenes.Address{
+						Ip:   yNode.address.Ip,
+						Port: yNode.address.Port,
+					},
+					Timeout: cleisthenes.DefaultDialTimeout,
+				})
+				if err != nil {
+					t.Fatalf("fail to dial to : %s, with error: %s", yNode.address, err)
+				}
+
+				go func() {
+					if err := conn.Start(); err != nil {
+						conn.Close()
+					}
+				}()
+
+				connPool.Add(yNode.address, conn)
+
+				member := cleisthenes.NewMember(yNode.address.Ip, yNode.address.Port)
+				memberMap.Add(member)
+			}
+		}
+
+		//
+		// set rbc
+		//
+		owner := memberMap.Member(mNode.address)
+		for _, member := range memberMap.Members() {
+			r := rbc.New(n, f, owner, member, connPool)
+			mNode.rbcMap[member.Address] = r
+		}
+
+		mNode.memberMap = memberMap
+		mNode.connPool = connPool
+	}
+
+	return nodeList
+}
+
+func RBCTest(proposer *mockNode, nodeList []*mockNode, data []byte) error {
+	iLogger.Infof(nil, "[START] proposer : %s propose data : %s", proposer.address.String(), data)
+
+	proposerMem := proposer.memberMap.Member(proposer.address)
+
+	if err := proposer.rbcMap[proposerMem.Address].HandleInput(data); err != nil {
+		return fmt.Errorf("error in MakeRequest : %s", err.Error())
+	}
+
+	doneCnt := 0
+	for {
+		for _, node := range nodeList {
+			value := node.rbcMap[proposerMem.Address].Value()
+			if value != nil {
+				if !bytes.Equal(value, data) {
+					return errors.New(fmt.Sprintf("[DONE] node : %s fail to reach an agreement - expected : %s, got : %s\n", node.address.String(), data, value))
+				} else {
+					iLogger.Infof(nil, "[DONE] node : %s has successfully reached an agreement ! value : %s\n", node.address.String(), value)
+				}
+				doneCnt++
+				break
+			}
+		}
+
+		if doneCnt == len(nodeList) {
+			break
+		}
+	}
+
+	return nil
+}
+
+func Test_RBC_4NODEs(t *testing.T) {
+	var n int = 4
+	var f int = 1
+
+	nodeList := setMockNodeList(t, n, f)
+	for i := 0; i < n; i++ {
+		data := fmt.Sprintf("node %d propose consensus", i)
+		if err := RBCTest(nodeList[i], nodeList, []byte(data)); err != nil {
```

Comment:
`RBCTest` seems that it waits one node to finish its task (shard -> interpolate) and starts next node.

Is it sequential test?

```
node1 start
wait...
node1 done

node2 start
wait...
node2 done

...
```

then I think we don't need `rbcMap` because node1,2,3,4 doesn't communicate **simultaneous but sequential**